forest/daemon/src/daemon.rs

```diff
@@ -538,37 +541,48 @@ async fn prompt_snapshot_or_die(
 async fn sync_from_snapshot<DB>(
     config: &Config,
     state_manager: &Arc<StateManager<DB>>,
+    proof_params_required: bool,
 ) -> Result<(), anyhow::Error>
 where
     DB: Store + Send + Clone + Sync + Blockstore + 'static,
 {
     if let Some(path) = &config.client.snapshot_path {
         let stopwatch = time::Instant::now();
+
         let validate_height = if config.client.snapshot {
             config.client.snapshot_height
         } else {
             Some(0)
         };
 
-        match import_chain::<_>(
+        let ensure_params_downloaded_task_handle = if proof_params_required {
+            Some(tokio::spawn(async {
+                // Add a small delay to avoid import chain progress bar being interrupted.
+                tokio::time::sleep(Duration::from_secs(1)).await;
+                forest_utils::proofs_api::paramfetch::ensure_params_downloaded().await
+            }))
+        } else {
+            None
+        };
+
+        import_chain::<_>(
             state_manager,
             &path.display().to_string(),
-            validate_height,
             config.client.skip_load,
         )
         .await
-        {
-            Ok(_) => {
-                info!("Imported snapshot in: {}s", stopwatch.elapsed().as_secs());
-            }
-            Err(err) => {
-                anyhow::bail!(
-                    "Failed miserably while importing chain from snapshot {}: {err}",
-                    path.display()
-                );
-            }
+        .context("Failed miserably while importing chain from snapshot")?;
+        info!("Imported snapshot in: {}s", stopwatch.elapsed().as_secs());
+
+        if let Some(ensure_params_downloaded_task_handle) = ensure_params_downloaded_task_handle {
+            ensure_params_downloaded_task_handle.await??;
+        }
+
+        if let Some(validate_height) = validate_height {
+            validate_chain(state_manager, validate_height).await?;
         }
```

Comment:
The `sync_from_snapshot` shouldn't re-validate the chain. You can hoist it out.

Reply:
As I understand, the whole point of this function is to combine the process of snapshot import and validation and make it cancellable. Would it be better to rename the function to `sync_from_snapshot_and_validate`?

```rust
tokio::select! {
        ret = sync_from_snapshot(&config, &state_manager, proof_params_required).fuse() => {
            if let Err(err) = ret {
                services.shutdown().await;
                return Err(err);
            }
        },
        _ = tokio::signal::ctrl_c() => {
            services.shutdown().await;
            return Ok(db);
        },
        _ = terminate.recv() => {
            services.shutdown().await;
            return Ok(db);
        },
        _ = shutdown_recv.recv() => {
            services.shutdown().await;
            return Ok(db);
        },
    }
```

Reply:
Hmm, the ctrl-c handling is bothering me to no end. It happens so often that ctrl-c is ignored. How about we move the ctrl-c handling out into `main.rs`? That way, the daemon doesn't have to worry about it.

Reply:
@lemmih When I search in the code, I can find 3 places that handle ctrl-c signal in forest daemon, I assume it's duplicated here for a reason.  @elmattic could u share ur insights on whether it is possible to remove the ctrl-c handling from here?

Reply:
There's no good reason to handle interrupts in three places. I'll have a go at refactoring it in a separate PR.

Reply:
> There's no good reason to handle interrupts in three places. I'll have a go at refactoring it in a separate PR.

@lemmih That makes sense, created an issue https://github.com/ChainSafe/forest/issues/2947 and mark this PR as blocked

Reply:
The interrupt code has been merged. Now we can continue to refactor the proof parameter handling.

Reply:
Done